servicetalk-encoding-netty/src/main/java/io/servicetalk/encoding/netty/NettyChannelContentCodec.java

```diff
@@ -0,0 +1,335 @@
+
+final class NettyChannelContentCodec extends AbstractContentCodec {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(NettyChannelContentCodec.class);
+    private static final Buffer END_OF_STREAM = DEFAULT_RO_ALLOCATOR.fromAscii(" ");
+
+    private final Supplier<MessageToByteEncoder<ByteBuf>> encoderSupplier;
+    private final Supplier<ByteToMessageDecoder> decoderSupplier;
+
+    NettyChannelContentCodec(final CharSequence name,
+                             final Supplier<MessageToByteEncoder<ByteBuf>> encoderSupplier,
+                             final Supplier<ByteToMessageDecoder> decoderSupplier) {
+        super(name);
+        this.encoderSupplier = encoderSupplier;
+        this.decoderSupplier = decoderSupplier;
+    }
+
+    @Override
+    public Buffer encode(final Buffer src, final int offset, final int length, final BufferAllocator allocator) {
+        requireNonNull(src);
+        final int availableBytes = src.readableBytes() - offset;
+        if (offset < 0) {
+            throw new IllegalArgumentException("Invalid offset: " + offset + " (expected >= 0)");
+        }
+
+        if (length <= 0) {
+            throw new IllegalArgumentException("Invalid length: " + length + " (expected > 0");
+        }
+
+        if (length > availableBytes) {
+            throw new IllegalStateException("Invalid length: " + length + " (expected <= " + availableBytes);
+        }
+
+        final MessageToByteEncoder<ByteBuf> encoder = encoderSupplier.get();
+        final EmbeddedChannel channel = newEmbeddedChannel(encoder, allocator);
+
+        try {
+            ByteBuf origin = toByteBuf(src, offset, length);
+            channel.writeOutbound(origin);
+
+            // May produce footer
+            preparePendingData(channel);
+            return requireNonNull(drainChannelOutToSingleBuffer(channel, allocator));
+        } catch (Throwable e) {
+            LOGGER.error("Error while encoding with {}", name(), e);
+            safeCleanup(channel);
+            throwException(e);
+            return EMPTY_BUFFER;
+        }
+    }
+
+    @Override
+    public Publisher<Buffer> encode(final Publisher<Buffer> from, final BufferAllocator allocator) {
+        requireNonNull(from);
+        return from
+                .concat(succeeded(END_OF_STREAM))
+                .liftSync(subscriber -> new PublisherSource.Subscriber<Buffer>() {
+
+                    private final MessageToByteEncoder<ByteBuf> encoder = encoderSupplier.get();
+                    private final EmbeddedChannel channel = newEmbeddedChannel(encoder, allocator);
+
+                    @Override
+                    public void onSubscribe(PublisherSource.Subscription subscription) {
+                        subscriber.onSubscribe(subscription);
+                    }
+
+                    @Override
+                    public void onNext(Buffer next) {
+                        if (next == null) {
+                            return;
```

Comment:
Should we request more?

Reply:
Propagating the NPE, this is not an expected item.